src/Badges.sol

```diff
@@ -115,6 +115,44 @@ contract Badges is
     specDataHolder = ISpecDataHolder(_dataHolder);
   }
 
+  function _give(
+    address _recipient,
+    string calldata _uri,
+    bytes calldata _signature,
+    uint256 _raftTokenId
+  ) internal returns (uint256) {
+    require(msg.sender != _recipient, "give: cannot give to self");
+    safeCheckAgreement(msg.sender, _recipient, _uri, _signature);
+    return mint(_recipient, _uri, _raftTokenId);
+  }
+
+  /**
+   * @notice Allows the owner of a badge spec to mint a badge to multiple recipeients who have requested it
+   * @param _recipients array the addresses who will receive a badge
+   * @param _uri the uri of the badge spec
+   * @param _signatures array of signatures that verify that the person receiving the badge actually requested it
+   */
+  function giveToMany(
+    address[] calldata _recipients,
+    string calldata _uri,
+    bytes[] calldata _signatures
+  ) external {
+    require(
+      _recipients.length == _signatures.length,
+      "giveToMany: recipients and signatures length mismatch"
+    );
+
+    uint256 raftTokenId = specDataHolder.getRaftTokenId(_uri);
+    require(
+      specDataHolder.getRaftOwner(raftTokenId) == msg.sender,
+      "giveToMany: unauthorized"
+    );
+
+    for (uint256 i = 0; i < _recipients.length; i++) {
```

Comment:
gas optimization opportunity

https://hackmd.io/@totomanov/gas-optimization-loops

Reply:
For minting 3 badges, putting the for loop inside of an `unchecked` takes `giveToMany` from 306213 to 306013 ... I'd say we hold off on `unchecked` for now until I research it further and feel more confident with it. (we can always upgrade)